app/models/migration_engine.rb

```diff
@@ -45,23 +56,76 @@ def base_product
   end
 
   # returns the possible migration targets for @installed_products by grouping successor products
-  def migration_targets
+  def migration_targets(migration_kind: :online)
+    migration_path_scope = (migration_kind == :online) ? ProductPredecessorAssociation.online : ProductPredecessorAssociation.offline
     installed_extensions = @installed_products.reject { |product| product == base_product }
-    base_successors = [base_product] + base_product.successors
-    extension_successors = installed_extensions.map { |e| [e] + e.successors }
-    # full set of combinations
-    combinations = base_successors.product(*extension_successors)
-    combinations.delete([base_product] + installed_extensions)
-    combinations
+    base_successors = base_product.successors.merge(migration_path_scope).to_a
+    base_successors.push(base_product) if migration_kind == :online
+    extension_successors = installed_extensions.map do |e|
+      successors = e.successors.merge(migration_path_scope)
+      successors += [e] if migration_kind == :online
+      successors
+    end
+    # full set of migrations
+    migrations = base_successors.product(*extension_successors)
+    migrations.delete([base_product] + installed_extensions)
+    migrations
   end
 
   # removes product combinations that include incompatible base-extension combinations
-  def remove_incompatible_combinations(migrations)
-    migrations.reject do |combination|
-      combination[1..-1].any? do |product|
-        (product.bases & combination).empty?
+  def remove_incompatible_migrations(migrations)
+    migrations.clone.each do |migration|
+      migration_root_product = migration.first
+      (migration - [migration_root_product]).each do |extension|
+        # remove migration if it does not include a valid root product for the current extension
+        if (extension.available_for.map(&:id) & migration.map(&:id)).empty?
+          migrations.delete(migration)
+          break
+        end
       end
     end
   end
 
+  # SLE 15 needs to have all of its available modules enabled, because there is
+  # no direct mapping from SLE12 modules to SLE15 modules.
+  # This method modifies the given array.
+  def add_dependencies(migrations)
+    migrations.map do |migration|
+      base = migration.first
+      migration.concat Product.available_or_recommended_modules(base.id) if base.version.split('.').first == '15'
```

Comment:
It seems like it doesn't account for whether the extension is mirrored or not.
I think the migration would fail should you have even one of the suggested modules not mirrored. :thinking:

Reply:
I suppose this calls for another card.